src/types/collectors.rs

```diff
@@ -0,0 +1,140 @@
+/*
+    Copyright © 2023, ParallelChain Lab
+    Licensed under the Apache License, Version 2.0: http://www.apache.org/licenses/LICENSE-2.0
+*/
+
+//! Definitions for the generic [SignedMessage] and [Collector] traits. Implementations used by the 
+//! [Pacemaker][crate::pacemaker::types] and [HotStuff][crate::hotstuff::types] protocols can be found 
+//! in the respective modules. The [Collectors] groups collectors for all active validator sets into
+//! a single struct, which can be easily updated on view or validator set state updates.
+
+pub use ed25519_dalek::{SigningKey, VerifyingKey, Signature};
+pub use sha2::Sha256 as CryptoHasher;
+
+use crate::messages::SignedMessage;
+use crate::state::{
+    block_tree::{BlockTree, BlockTreeError}, 
+    kv_store::KVStore
+};
+
+use super::{
+    basic::{ChainID, TotalPower, ViewNumber}, 
+    validators::{ValidatorSet, ValidatorSetState}
+};
+
+/// Evidence that a quorum of validators from a given validator set supports a given decision. The 
+/// evidence comes in the form of a set of signatures of the validators.
+pub trait Certificate {
```

Comment:
What kind of semantics should implementations of this trait give for `is_correct` and `is_correctly_signed`? E.g., should `is_correctly_signed` also check whether the "Certificate" is signed by a quorum of validators? And should `is_correct` imply `is_correctly_signed`?